src/terminator/intrinsic.rs

```diff
@@ -299,35 +268,35 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             }
 
             "min_align_of" => {
-                let elem_ty = substs.type_at(0);
+                let elem_ty = instance.substs.type_at(0);
                 let elem_align = self.type_align(elem_ty)?;
                 let align_val = PrimVal::from_u128(elem_align as u128);
                 self.write_primval(dest, align_val, dest_ty)?;
             }
 
             "pref_align_of" => {
-                let ty = substs.type_at(0);
+                let ty = instance.substs.type_at(0);
                 let layout = self.type_layout(ty)?;
                 let align = layout.align(&self.tcx.data_layout).pref();
                 let align_val = PrimVal::from_u128(align as u128);
                 self.write_primval(dest, align_val, dest_ty)?;
             }
 
             "move_val_init" => {
-                let ty = substs.type_at(0);
+                let ty = instance.substs.type_at(0);
                 let ptr = arg_vals[0].read_ptr(&self.memory)?;
                 self.write_value_to_ptr(arg_vals[1], ptr, ty)?;
             }
 
             "needs_drop" => {
-                let ty = substs.type_at(0);
+                let ty = instance.substs.type_at(0);
                 let env = self.tcx.empty_parameter_environment();
                 let needs_drop = self.tcx.type_needs_drop_given_env(ty, &env);
                 self.write_primval(dest, PrimVal::from_bool(needs_drop), dest_ty)?;
             }
 
             "offset" => {
-                let pointee_ty = substs.type_at(0);
+                let pointee_ty = instance.substs.type_at(0);
```

Comment:
Do you still plan to do some cleanups? Should I wait until those are done before reviewing?